libbeat/reader/readjson/json.go

```diff
@@ -127,41 +141,47 @@ func createJSONError(message string) common.MapStr {
 	return common.MapStr{"message": message, "type": "json"}
 }
 
-func (pp *JSONPostProcessor) Name() string {
-	return "json"
-}
+// Next decodes JSON and returns the filled Line object.
+func (p *JSONParser) Next() (reader.Message, error) {
+	fieldsKey := "json" + p.id
+	message, err := p.JSONReader.nextWithKey(fieldsKey)
+	if err != nil {
+		return message, err
+	}
 
-func (pp *JSONPostProcessor) PostProcess(msg *reader.Message) {
-	jsonFields, ok := msg.Fields[pp.Name()].(common.MapStr)
+	jsonFields, ok := message.Fields[fieldsKey].(common.MapStr)
 	if !ok {
-		return
+		return message, err
 	}
 
 	// The message key might have been modified by multiline
-	if len(pp.cfg.MessageKey) > 0 && len(msg.Content) > 0 {
-		jsonFields[pp.cfg.MessageKey] = string(msg.Content)
+	if len(p.cfg.MessageKey) > 0 && len(message.Content) > 0 {
+		jsonFields[p.cfg.MessageKey] = string(message.Content)
 	}
 
 	// handle the case in which r.cfg.AddErrorKey is set and len(jsonFields) == 1
 	// and only thing it contains is `error` key due to error in json decoding
 	// which results in loss of message key in the main beat event
 	if len(jsonFields) == 1 && jsonFields["error"] != nil {
-		msg.Fields["message"] = string(msg.Content)
+		message.Fields["message"] = string(message.Content)
 	}
 
-	var id string
-	if key := pp.cfg.DocumentID; key != "" {
+	if key := p.cfg.DocumentID; key != "" {
 		if tmp, err := jsonFields.GetValue(key); err == nil {
-			if v, ok := tmp.(string); ok {
-				id = v
+			if id, ok := tmp.(string); ok {
 				jsonFields.Delete(key)
+
+				if message.Meta == nil {
+					message.Meta = common.MapStr{}
+				}
+				message.Meta["_id"] = id
 			}
 		}
 	}
 
-	if pp.cfg.KeysUnderRoot {
+	if p.cfg.KeysUnderRoot {
 		// Delete existing json key
-		delete(msg.Fields, "json")
+		delete(message.Fields, fieldsKey)
```

Comment:
This post processing get + merge + delete step is not really required. As we do all changes from within `Next`, we can keep all fields to be merged into the event in a separate `MapStr`, massage the message if there is need, and finally add the temporary `MapStr` into `message.Fields`.